chapters/appendix-IV-SPDX-license-expressions.md

```diff
@@ -1,175 +1,174 @@
 
-The exact syntax of license expressions is described below in [ABNF](http://tools.ietf.org/html/rfc5234).
+## IV.2 Syntax <a name="IV.2"></a>
 
-idstring = 1*(ALPHA / DIGIT / "-" / "." )
+The exact syntax of license expressions is described below in [ABNF][rfc5234].  `ALPHA`, `DIGIT`, and `WSP` are from the [ABNF core rules][rfc5234-aB].
 
-license-id = \<short form license identifier in [Appendix I.1](./appendix-I-SPDX-license-list.md)>
+```
+idstring                    = 1*(ALPHA / DIGIT / "-" / "." )
+license-id                  = <short form license identifier>
+license-exception-id        = <short form license exception identifier>
+license-ref                 = ["DocumentRef-" idstring ":"] "LicenseRef-" idstring
+license                     = license-id / license-ref
+space                       = 1*WSP
+simple-expression           = license ["+"] [space "WITH" space license-exception-id]
+binary-compound-operator    = "AND" / "OR"
+license-expression          = simple-expression
+                            / license-expression space binary-license-operator space license-expression
+                            / "(" license-expression ")"
+                            / space license-expression
+                            / license-expression space
+enclosed-license-expression = [space] simple-expression [space]
+                            / [space] "(" [space] license-expression [space] ")" [space]
+```
 
-license-exception-id = \<short form license exception identifier in [Appendix I.2](appendix-II-license-matching-guidelines-and-templates.md)>
+Where:
 
-license-ref = ["DocumentRef-"1\*(idstring)":"]"LicenseRef-"1*(idstring)
+* Short form license identifiers are drawn from sections [I.1](appendix-I-SPDX-license-list.md#I.1) and [I.3](appendix-I-SPDX-license-list.md#I.3).
+* Short form license exception identifiers are drawn from [section I.2](appendix-I-SPDX-license-list.md#I.2).
 
-simple-expression = license-id / license-id”+” / license-ref
+When a particular rule is not referenced, “SPDX License Expressions” means the more general `license-expression`.
+Consumers that need the more limited `enclosed-license-expression` (which may be easier to parse depending on the surrounding context) SHOULD say so explicitly.
 
-compound-expression = 1*1(simple-expression /
+Tag:value authors SHOULD use `enclosed-license-expression` instead of the more general `license-expression`.
```

Comment:
SHOULD -> MUST - this is a rule we agreed to to enable machine parsing

Reply:
> SHOULD -> MUST - this is a rule we agreed to to enable machine parsing

But 2.1 [uses][1]:

> For the Tag:value format, any license expression that consists of more than one license identifier and/or LicenseRef, should be encapsulated by parentheses

I think making this stricter without a major version bump is a bit dicey, especially with [wild exceptions that don't set those enclosing parens][2].

On the other hand, [the tag:value format is currently poorly spec'ed][3], so I guess we can always tell broken docs and tools “sorry, but you should have known you were making assumptions about the eventual format once it was spec'ed out”.

[1]: https://github.com/spdx/spdx-spec/blame/cfa1b9d08903befdf03e669da6472707b7b60cb9/chapters/appendix-IV-SPDX-license-expressions.md#L62
[2]: https://github.com/spdx/spdx-spec/issues/45#issuecomment-336303074
[3]: https://github.com/spdx/spdx-spec/issues/22#issuecomment-327253273

Reply:
@wking Good point - I would like to "correct" the 2.1 spec.  

This was highly debated at the time, but I believe there was a consensus to require the parenthesis.  I do know that my parser would break without the parenthesis and I believe the syntax is ambiguous making it unreliable for any implementation of a parser for tag/value  @kestewart - let me know if you agree we can change this to "must".

Reply:
> I do know that my parser would break without the parenthesis and I believe the syntax is ambiguous making it unreliable for any implementation of a parser for tag/value…

Parsers can be patched, but the data they're parsing may be historical (like the wild example I link above).  I'm interested in having *new* parsers be able to parse as many *existing* declarations as is reasonably possible.  I'm less worried about breaking existing parsers.

Reply:
@wking This was discussed and resolved during 2.1 development that tag/value files need parenthesis if the expression is complex.  I consider it a defect in the spec that this wasn't worded as a must.  The reference to parsers is just pointing out the implications of not documenting what was agreed to.

Since you disagree, why don't you keep it as should and I'll create my own pull request after yours is complete to change it to must.  That way it will not appear that you are requesting a change that you disagree with.  We can then pick up the debate on the new pull request.

Reply:
IMHO In most cases parens are not needed and just make the expression heavier and ugly looking for no good reason. I think that rather than mandating them, we should relax their use in upcoming spec versions. 

Also in the common case it never makes sense to have multi-line expressions .... so we should IMHO state that expression MUST be a single line. In doing so you make things cleaner and simpler.

Reply:
> IMHO In most cases parens are not needed and just make the expression heavier and ugly looking for no good reason.

“We want an unambiguous way to identify the end of this value” is a good reason.  Requiring (or at least strongly recommending) enclosing parens is one way to do that.  Defining clear folding-whitespace rules (#44, #45) is another way.  Forbidding newlines in values is a third way.  All of those approaches have pros and cons, especially in the context of (backwards) compatibility.

And debating this particular issue was not a priority for this PR.  Just tell me what I need to say here to get this landed, and we can come back and kick this particular point around in follow-up work (e.g. #44 and #45).